### PR TITLE
Add support for reversible suspensions through ActivityPub 

### PR DESCRIPTION
**Pull request by Gargron**
Follow-up to #14726

- Change reversibly suspended accounts to return empty data instead of HTTP 410 (while those that cannot be restored will still return HTTP 410)
- Change Webfinger resolution to queue full account deletion when it encounters a HTTP 410 error for an account that's still in the database

Suspension is expressed through boolean `suspended` attribute on the actor, and changes are distributed using the standard `Update` activity. This means that handling suspended accounts works pretty much the same if you receive the event live or encounter the suspended account in the wild. Accounts that were suspended locally cannot be unsuspended remotely. On the other hand, we no longer ignore `Update`, `Undo`, `Reject`, and `Delete` events for accounts that were suspended.

### CODE REVIEW

File: app/models/account.rb
Comment (ThibG) [inline app/models/account.rb:227]: Slightly worried this logic could fall apart if at some point we (or an admin) end up deciding to cancel some deletion requests without unsuspending the user, but otherwise it seems fine.

File: app/serializers/activitypub/actor_serializer.rb
Comment (ThibG) [inline app/serializers/activitypub/actor_serializer.rb:159]: I wonder if these could cause subtle issues in `Move` handling or whatnot... but it's unlikely.

File: app/services/activitypub/process_account_service.rb
Comment (ThibG) [inline app/services/activitypub/process_account_service.rb:106]: Doesn't this mean that in the following scenario, you'll miss the account being suspended?
- you locally suspend a remote account
- the remote account gets suspended on their instance
- you locally unsuspend that account

File: app/services/activitypub/process_collection_service.rb
Comment (ThibG) [inline app/services/activitypub/process_collection_service.rb:36]: I would allow `Delete` on any object, as well as `Undo Follow` / `Reject Follow` and that kind of stuff, to be honest.

Rationale for `Delete`:
- remote admin suspends their user
- then remove offending posts (while the user is suspended)
- undoes the suspension

Rationale for `Undo Follow` / `Reject Follow`:
- better handling of Unfollow/Refollow flow as suggested in #14775

File: db/post_migrate/20201017234926_fill_account_suspension_origin.rb
Comment (ThibG) [inline db/post_migrate/20201017234926_fill_account_suspension_origin.rb:7]: should not matter much, but I would add a `.where(suspension_origin: nil)` here

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
I'll have to do a second review pass, but here are a few inline comments (most of them minor) for now.

More generally, I'm a bit worried about the `suspension_origin` enum making the logic a bit more complicated and error-prone (but I also understand it's useful for not having to change how suspended accounts are looked up/checked).

Another thing I'm a bit worried about is the two instances not agreeing on the suspension status of some user, which could result in yet another occurrence of “ghost followers”.

File: app/services/activitypub/process_account_service.rb
Comment (ThibG) [reply app/services/activitypub/process_account_service.rb:106]: I think this could be solved by doing the following when unsuspending a locally-suspended remote account:
1. change the suspension origin from local to remote (even if the account is not suspended remotely)
2. refresh the account profile (if the account is not suspended remotely, it will be marked as unsuspended)

File: app/controllers/concerns/account_controller_concern.rb
Comment (ThibG) [inline app/controllers/concerns/account_controller_concern.rb:55]: I am slightly worried this could cause some implementations to entirely delete information about the account. We entertained the idea of deleting something on 410 before, and isn't it what `tootctl accounts cull` would do?

**Review by ClearlyClaire — COMMENTED** [review]
I will do another review pass later today, but it seems mostly good. So far, my only worry is about returning 410 for temporary suspensions, as that might trigger deletion in some implementations.

File: app/controllers/concerns/account_controller_concern.rb
Comment (Gargron) [reply app/controllers/concerns/account_controller_concern.rb:55]: Yeah but I think it should be fine as long as the json-ld variant returns HTTP 200...

Comment (Gargron) [reply app/controllers/concerns/account_controller_concern.rb:55]: Speaking of which, I need to make that one use DeleteAccountService instead of SuspendAccountService

File: app/controllers/settings/deletes_controller.rb
Comment (ThibG) [inline app/controllers/settings/deletes_controller.rb:45]: Hmm thinking about that, a self-deletion being displayed as a suspension might be misleading and problematic. Though that's a permanent suspension so I guess it should be fine, just returning 410 with a generic “Gone” page.

File: app/services/resolve_account_service.rb
Comment (ThibG) [inline app/services/resolve_account_service.rb:97]: Note that we currently return `410 Gone` in webfinger for suspended accounts, regardless of whether the suspension is final: https://github.com/tootsuite/mastodon/blob/acc1c038610a82d8470acde6753716da0dc4b1fa/app/controllers/well_known/webfinger_controller.rb#L37-L39

It should at the very least be changed for this PR.

File: app/controllers/concerns/account_controller_concern.rb
Comment (ThibG) [reply app/controllers/concerns/account_controller_concern.rb:55]: I'm still uneasy with returning 410 here, but 200 doesn't seem super fitting either… just checked Twitter and it returns 200 for suspended accounts…

Likewise, RSS, is, I think, used by Hubzilla, and I would worry about it deciding to delete accounts on 410 (I don't know if it does, I haven't checked).

File: app/controllers/concerns/account_owned_concern.rb
Comment (ThibG) [inline app/controllers/concerns/account_owned_concern.rb:46]: Same here, I'm uneasy with `410 Gone`.

File: app/services/activitypub/process_collection_service.rb
Comment (ThibG) [inline app/services/activitypub/process_collection_service.rb:36]: We probably want to allow deletion of content even for locally-suspended remote accounts here. Two reasons:
- suspending a remote account that then gets deleted on the remote server would make us miss the `Delete`
- we should let the remote suspended account still remove followers (we clear the follow relationships from the suspended account, not the other way around)

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
In addition to the inline comments:
- `DeleteAccountService#purge_profile!` needs to set suspension origin
- `BlockDomainService#suspend_accounts!` and `BlockDomainService#process_retroactive_updates!` need to (un)set suspension origin
- `ActivityPub::ProcessAccountService#create_account` needs to set suspension origin on auto_block

Also, I think we could/should federate silencing information as well, this would make some things less hackish. This definitely can be in another PR though.

**Comment by Gargron** [comment]
I have removed the "suspended" profile design from the PR. Temporarily suspended profiles return HTTP 403.

**Review by ClearlyClaire — APPROVED** [review]
The logic behind which response to return in case of suspension is a bit hard to follow but I don't really have a better idea on how to implement it right now.

Also, I'm not sure why you removed the special page about the account being suspended, it does seem helpful to me. In any case that can be added back in another PR.

I'd like to see some actual testing of the feature (I might do that later), but overall, the PR looks good to me.

File: app/services/unsuspend_account_service.rb
Comment (ThibG) [inline app/services/unsuspend_account_service.rb:41]: Note that `merge_into_timeline` does not exist. This is a bug from the latest FeedManager refactoring, In a similar way, `unmerge_from_timeline` does not exist either. (#15099)

File: app/services/suspend_account_service.rb
Comment (ThibG) [inline app/services/suspend_account_service.rb:37]: This should remove from followers on the remote end (although for some reason it doesn't appear to work?), but this doesn't remove it from the local followers list. In the `DeleteAccountService`, that wasn't an issue since we deleted every relationship right away, but here they are kept around.

**Pull request merged**
